backend/controllers/authController.js

```diff
@@ -35,21 +58,29 @@ exports.register = async (req, res) => {
         // Send response
         res.status(201).json({ message: 'User created' });
     } catch (err) {
-        res.status(500).json({ error: err.message });
+        res.status(500).json({ message: 'Registration failed' });
     }
 };
 
 exports.login = async (req, res) => {
     try {
         const { identifier, password } = req.body; // Identifier can either be username or email
-        // Need to add data validation
-        //
-        //
+        
+        if (!identifier || !password) {
+            return res.status(400).json({ message: 'Please provide both identifier and password' });
+        }
+
+        let query;
+        if (isValidEmail(identifier)) {
+            query = { email: identifier };
+        } else if (isValidUsername(identifier)) {
+            query = { username: identifier };
+        } else {
+            return res.status(400).json({ message: 'Invalid username or email format' });
```

Comment:
Do we have to send an error for invalid format, since the server would return an invalid credentials error anyway?